model/diffusion/diffusion_ppo_sumlikelihood.py

```diff
@@ -0,0 +1,220 @@
+"""
+DPPO: Diffusion Policy Policy Optimization with summed likelihood. 
+
+K: number of denoising steps
+To: observation sequence length
+Ta: action chunk size
+Do: observation dimension
+Da: action dimension
+
+C: image channels
+H, W: image height and width
+
+"""
+
+from typing import Optional
+import torch
+import logging
+import math
+
+log = logging.getLogger(__name__)
+from model.diffusion.diffusion_vpg import VPGDiffusion
+
+
+class PPODiffusionSumLikelihood(VPGDiffusion):
+    def __init__(
+        self,
+        gamma_denoising: float,
+        clip_ploss_coef: float,
+        clip_ploss_coef_base: float = 1e-3,
+        clip_ploss_coef_rate: float = 3,
+        clip_vloss_coef: Optional[float] = None,
+        clip_advantage_lower_quantile: float = 0,
+        clip_advantage_upper_quantile: float = 1,
+        norm_adv: bool = True,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        # Whether to normalize advantages within batch
+        self.norm_adv = norm_adv
+
+        # Clipping value for policy loss
+        self.clip_ploss_coef = clip_ploss_coef
+        self.clip_ploss_coef_base = clip_ploss_coef_base
+        self.clip_ploss_coef_rate = clip_ploss_coef_rate
+
+        # Clipping value for value loss
+        self.clip_vloss_coef = clip_vloss_coef
+
+        # Discount factor for diffusion MDP
+        self.gamma_denoising = gamma_denoising
+
+        # Quantiles for clipping advantages
+        self.clip_advantage_lower_quantile = clip_advantage_lower_quantile
+        self.clip_advantage_upper_quantile = clip_advantage_upper_quantile
+
+    def loss(
+        self,
+        obs,
+        chains_prev,
+        chains_next,
+        denoising_inds,
+        returns,
+        oldvalues,
+        advantages,
+        oldlogprobs,
+        use_bc_loss=False,
+        reward_horizon=4,
+    ):
+        """
+        PPO loss
+
+        obs: dict with key state/rgb; more recent obs at the end
+            state: (B, To, Do)
+            rgb: (B, To, C, H, W)
+        chains: (B, K+1, Ta, Da)
+        returns: (B, )
+        values: (B, )
+        advantages: (B,)
+        oldlogprobs: (B, K, Ta, Da)
+        use_bc_loss: whether to add BC regularization loss
+        reward_horizon: action horizon that backpropagates gradient
+        """
+        # repeat the obs for each denoising step
+        B = chains_next.shape[0]
+        obs_repeat = {
+            "state": obs["state"].repeat_interleave(self.ft_denoising_steps, dim=0)
+        }
+        denoising_inds = denoising_inds.repeat_interleave(chains_prev.shape[0])
```

Comment:
This is wrong? `obs_repeat` now has obs_1, obs_1, obs_1, ... obs_2, obs_2,...., but denoising_inds has 0, 0,0,0,0,,,,...1,1,1,1,1, .....

I think you want repeat instead of repeat_interleave for denoising_inds